src/bin/str.rs

```diff
@@ -89,88 +132,92 @@ define_language! {
         "+" = Add([Id; 2]),
         "-" = Sub([Id; 2]),
         Lit(Constant),
-        Var(egg::Symbol),
     }
 }
 
-/// Concatenate strings.
-fn concat(a: &[u8], b: &[u8]) -> SmallStr {
-    a.iter().chain(b).copied().collect()
-}
-
 impl SynthLanguage for Lang {
     type Constant = Constant;
     type Type = Type;
 
     fn get_type(&self) -> Self::Type {
-        Type::Top
+        match self {
+            Lang::Lit(c) => match c {
+                Constant::Num(_) => Type::Number,
+                Constant::Str(_) => Type::String,
+            },
+            Lang::SVar(_)
+            | Lang::Concat(_)
+            | Lang::Replace(_)
+            | Lang::At(_)
+            | Lang::SubStr(_)
+            | Lang::FromInt(_) => Type::String,
+            Lang::NVar(_)
+            | Lang::Len(_)
+            | Lang::ToInt(_)
+            | Lang::IndexOf(_)
+            | Lang::Add(_)
+            | Lang::Sub(_) => Type::Number,
+        }
     }
 
     fn eval<'a, F>(&'a self, cvec_len: usize, mut v: F) -> CVec<Self>
     where
         F: FnMut(&'a Id) -> &'a CVec<Self>,
     {
         match self {
-            Lang::Concat([a, b]) => map!(v, a, b => {
-                let a = a.to_slice().unwrap();
-                let b = b.to_slice().unwrap();
-                Some(Constant::Str(concat(a, b)))
-            }),
+            Lang::SVar(_) => vec![],
+            Lang::NVar(_) => vec![],
+            Lang::Lit(c) => vec![Some(c.clone()); cvec_len],
+            Lang::Concat([a, b]) => {
+                map!(v, a, b => Some(Constant::Str(SmallString::from_str(&format!("{}{}", a.to_str().unwrap(), b.to_str().unwrap())))))
+            }
             Lang::Replace([s, t, t2]) => map!(v, s, t, t2 => {
                 let s = s.to_str().unwrap();
                 let t = t.to_str().unwrap();
                 let t2 = t2.to_str().unwrap();
                 let out = s.replacen(t, t2, 1);
-                Some(Constant::Str(out.into_bytes().into()))
+                Some(Constant::Str(SmallString::from_str(&out)))
             }),
             Lang::At([a, i]) => map!(v, a, i => {
-                let a = a.to_slice().unwrap();
+                let a = a.to_smallstr().unwrap();
                 let i = i.to_num().unwrap();
-                if 0 <= i && i < (a.len() as _) {
-                    let i = i as usize;
-                    Some(Constant::Str(a[i..i+1].into()))
+                if i < 0 || i > (a.len() as _) {
```

Comment:
shouldn't this be `i >= (a.len() as _)` ?

Reply:
Yep, you're right, my bad!